### PR TITLE
Added the ability to assign the computer's local hostname to id and get the computer's local ip address

### PR DESCRIPTION
**Pull request by rdash99**
Examples have been created to demonstrate - however, as node.py needs to be modified to accept ID as an attribute I have placed the modified copy in the examples folder as the unmodified package is currently used by default if the import references p2pnetwork.node.

I decided to make these minor changes and contribute this as I needed the ability to assign an ID as well as automatically assign a local IP to a node, and although the ID may not always be unique, it should serve well enough to identify a locally networked machine in a human-readable manner which is potentially useful when built into an application with a GUI interface.

I hope this is useful to someone in the future.

### CODE REVIEW

**Comment by macsnoeren** [comment]
I will check it shortly :).

**Comment by cgsops** [comment]
is this can run behind nat?

**Comment by macsnoeren** [comment]
Note that this is a framework offcourse, so you can implement a peer-to-peer application. The id of the nodes are self generated, see node.py (line 62-65). Furthermore, the nodeconnection.py holds a key-value data store to store meta information about the node, that can be a local IP for instance. This can be accessed using:

'''python
def set_info(self, key, value):        self.info[key] = value
def get_info(self, key):        return self.info[key
'''

Is this already good for you? 

However, I can imagine that it is very nice when it is possible to set the ID of the node yourself. For example when it is constructed. The datastore can be used to store the local IP and other information. I would like to follow this approach. Is that okay for you?

File: examples/my_own_p2p_application hostname example.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+# get the local computer hostname:
+id = socket.gethostname()
+s.close()
+
+# id is added as a third argument and ip can be used to replace the first argument:
+node_1 = MyOwnPeer2PeerNode("127.0.0.1", 8001, id)
+node_2 = MyOwnPeer2PeerNode("127.0.0.1", 8002, id)
+node_3 = MyOwnPeer2PeerNode("127.0.0.1", 8003, id)
+
```
Comment (macsnoeren) [inline examples/my_own_p2p_application hostname example.py:33]: De nodes need to have unique ID's.

File: examples/node.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+      port: The port number that is used to bind the TCP/IP server to.
+      callback: (optional) The callback that is invokes when events happen inside the network
+               def node_callback(event, main_node, connected_node, data):
+                 event: The event string that has happened.
+                 main_node: The main node that is running all the connections with the other nodes.
+                 connected_node: Which connected node caused the event.
+                 data: The data that is send by the connected node."""
+
+    def __init__(self, host, port, id, callback=None):
+        """Create instance of a Node. If you want to implement the Node functionality with a callback, you should
```
Comment (macsnoeren) [inline examples/node.py:37]: You can modify the real class. Why did you create a new file?

File: examples/node.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+      host: The host name or ip address that is used to bind the TCP/IP server to.
+      port: The port number that is used to bind the TCP/IP server to.
+      callback: (optional) The callback that is invokes when events happen inside the network
+               def node_callback(event, main_node, connected_node, data):
+                 event: The event string that has happened.
+                 main_node: The main node that is running all the connections with the other nodes.
+                 connected_node: Which connected node caused the event.
+                 data: The data that is send by the connected node."""
+
+    def __init__(self, host, port, id, callback=None):
```
Comment (macsnoeren) [inline examples/node.py:36]: When the id is not used, we could fall back on the automatic generation of the id as we had. Is that okay?

**Review by macsnoeren — CHANGES_REQUESTED** [review]
Check my comments ;)

**Comment by SamuelHaidu** [comment]
I think that we need to remove the id handshake in the base node class. This functionality need to be implemented by programmer because is in the level of aplication not in the level of network. This can be implemented in a child class like the examples.

**Comment by macsnoeren** [comment]
I agree, so what I proposed is okay to implement? for testing purposes, the generated id would be okay to put in. However, I agree to the fact the the application should deal with this: Only unique ID's should exist in the network. Could you provide some answers to my questions? Thanks!

**Comment by macsnoeren** [comment]
> is this can run behind nat?

Yes and no. A node is able to recieve connections from other nodes and make connections to other nodes. That is the principle of a peer-to-peer network. So, it cannot run behind nat. However, when you have a node that only needs to connect to another node(s) (some kind of leaf node), you can run it behind a nat. So you alway need nodes that are accessible.

**Comment by SamuelHaidu** [comment]
> I agree, so what I proposed is okay to implement? for testing purposes, the generated id would be okay to put in. However, I agree to the fact the the application should deal with this: Only unique ID's should exist in the network. Could you provide some answers to my questions? Thanks!

Yes, is a good idea that every node has a unique ID (may be change the 512 bits hash to UUID, what you think?) but only locally. Check lines 216/217 that node exchange over network your ids. This implies that a application implemented with this framework has a little protocol. I think that in this moment the id is not useful to send messages. The id is only a attribute in NodeConnection object. I think the changes you made to the code don't change much in practice.

**Comment by macsnoeren** [comment]
I will check the UUID. It is a good idea.

**Pull request merged**